Remove default constructor from SecretKey class

ckeshava opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## High Level Overview of Change

This Pull Request deletes the default constructor from SecretKey and PublicKey classes. 

All the constructed Public Keys are valid, non-empty and contain 33 bytes of data. This PR establishes this invariant.

Additionally, the memory footprint of the `PublicKey` class is reduced. The `size_` data member is declared as `static`. 



<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

Fix https://github.com/XRPLF/rippled/issues/2942

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

ckeshava commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -802,7 +802,7 @@ RCLConsensus::Adaptor::validate(
         lastValidationTime_,
         validatorKeys_.publicKey,
-        validatorKeys_.secretKey,
+        *validatorKeys_.secretKey,
```
I'm not sure if this is the correct way to solve this problem. I'm assuming that `ValidatorKeys::SecretKey` is set before this line is executed.

The said optional is constructed inside the `ValidatorKeys` constructor at https://github.com/XRPLF/rippled/blob/66bfe909e6de0dd1c7be0d9c01e088d3fed7f0f3/src/ripple/app/misc/impl/ValidatorKeys.cpp#L29

ckeshava commented:
@nbougalis  I was able to eliminate the `PublicKey::size_` data member !
we had discussed this idea one year ago during my summer internship :))

ckeshava commented:
@scottschurr Can you please take a look at this PR whenever you are free? Thanks!

scottschurr commented on src/ripple/app/misc/ValidatorKeys.h:
```diff
@@ -36,13 +36,14 @@ class Config;
-    PublicKey masterPublicKey;
-    PublicKey publicKey;
-    SecretKey secretKey;
+    std::optional<PublicKey> masterPublicKey;
+    std::optional<PublicKey> publicKey;
+    std::optional<SecretKey> secretKey;
```
As soon as I started reviewing this pull request I was stumbling across these `optional`s in RCLConsensus.cpp.  I got really uncomfortable because this new interface suggests that any arbitrary combination of the keys may be missing.  There's a lot of validation that needs to happen to use these safely.

I struggled looking for a solution for several hours and went down a couple of dead ends.  But along the way I found out that either _all_ of these keys are valid or _none_ of them are.  That eventually led to this solution.  Suppose you declare `ValidatorKeys` like this:
```
class ValidatorKeys
{
public:
    // Make a struct to group the keys.  Either all keys are valid or none are.
    struct Keys
    {
        PublicKey masterPublicKey;
        PublicKey publicKey;
        SecretKey secretKey;

        Keys() = delete;
        Keys(
            PublicKey const& masterPublic_,
            PublicKey const& public_,
            SecretKey const& secret_)
        : masterPublicKey(masterPublic_)
        , publicKey(public_)
        , secretKey(secret_)
        { }
    };

    std::optional<Keys> keys;
    NodeID nodeID;
    std::string manifest;
    std::uint32_t sequence = 0;

    ValidatorKeys() = delete;
    ValidatorKeys(Config const& config, beast::Journal j);

    bool
    configInvalid() const
    {
        return configInvalid_;
    }

private:
    bool configInvalid_ = false;  //< Set to true if config was invalid
};
```
Now there's only one `optional`: `keys`.  If `keys` has a value, then I can use any and all of the keys it holds.  Only one place to check to validate using any of them.

I tried to shorten the names of the members of `Keys`.  But I was not successful.  There are three obvious shorter names:

- `masterPublic`
- `public`, and
- `secret`.

Unfortunately for this approach, `public` is a reserved word in C++.  So that won't work.  Oh well.

If you use this suggestion, be sure to use `std::optional::emplace()` to initialize `keys` in the constructor.

Thoughts?

ckeshava replied:
thank you Scott, I concur with your approach. I had a similar apprehension about the optionals, but I wasn't able to figure out a better solution. I'll work on this approach.

Why do you emphasize on the use of `std::optional::emplace()` ? Is there a performance benefit from the direct-initialization (apart from perfect forwarding) ?

scottschurr replied:
Using `emplace()` is not a huge deal, but it's easy to miss and may provide a minor optimization.

In your `ValidatorKeys` constructor from a `Config` you need to initialize `keys`.  With `emplace()` it looks like this:
```
                keys.emplace(m->masterKey, pk, token->validationSecret);
```
Without `emplace()` it might look like this:
```
                keys = Keys(m->masterKey, pk, token->validationSecret);
```
That results in constructing one more `Keys` object than when `emplace()` is used.  That, in turn, involves copying each key an extra time.  Not a killer, but why do the extra copies if they are not needed?

ckeshava replied:
okay 👍 
I read that `emplace` constructs objects "in-place" (Docs: https://en.cppreference.com/w/cpp/utility/optional/emplace). Are we constructing a `Keys` object and moving it into the optional's location? I'm trying to understand how the in-place construction occurs (especially for classes with inheritance)

scottschurr replied:
A `std::optional` can be thought of as containing two members:

1. A `bool` that indicates whether the `optional` has a value.
2. A C-style array of `std::byte` that has the proper alignment and adequate size to hold the type contained by the `optional`.

When an `optional` is default constructed the `bool` is set to `false` and the C-style array is left uninitialized.

When an empty optional is assigned to, the copy or move constructor is used to copy or move an already constructed contained type into the C-style array owned by the `optional`.  Once that move or copy is complete the `bool` is set to `true`.  Notice that there are two constructor calls: 1) build the type you want outside of the `optional` 2) construct that same type into the C-style array inside the `optional`.

When you call `emplace()` on an empty optional, then the type being emplaced in the optional does not need to previously exist.  Instead, `emplace()` does a "placement new" of the contained type into the C-style array owned by the `optional`.  We only need to call the constructor for the emplaced type once.

Make sense?  Questions?

scottschurr reviewed (changes requested):
You got half of the point of my previous comment.  But I was not clear enough about the motivation.  Sorry about that.

Every time code is about to access to contents of an `optional` it needs to do one of two things:

- Have some prior knowledge so it "knows" that the `optional` has valid contents, or
- Verify, locally, that the `optional` has a valid value prior to access.

If the code does not do one of those two things then that's a setup for a crash in the future.

The safest choice of those two, particularly in the face of an evolving code base, is to do the local check.

Putting all of the `*Key`s in `ValidatorKeys` into a single `optional` means that with one check...
```
if (validatorKeys_.keys)
```
I can be assured that it's safe to access any and all of the `*Key`s in `ValidatorKeys::Keys`.

If the `*Key`s are stored in three separate `optional`s, then I have to do separate checks any time I access any of the keys.

So, by putting all three `*Key`s in a single `optional` it becomes both

1. Easy and
2. Compute efficient
 
to see whether accessing _any_ of the `*Key`s is safe.

Just in case I'm still not clear, I've made a commit where the `optional` is checked _every time_ before the `*Key`s in `ValidatorKeys` are accessed.  That's the top-most commit here: https://github.com/scottschurr/rippled/commits/chenna-ppp

You're welcome to cherry-pick that commit if you want.  But don't blindly drop it in.  Look at the places the checks are added.  I might have messed something up.

Does this all make sense?  From a larger perspective we're pushing first for safety and correctness.  Once we have that, then we can make things efficient.

ckeshava commented:
Okay, that clarifies things, thank you. I had my reservations about the
getter functions, I did not think about the downstream effects of an
unseated optional.

I'll make the necessary changes :+1

Regards,
Keshava


On Thu, Aug 3, 2023 at 2:19 PM Scott Schurr ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> You got half of the point of my previous comment. But I was not clear
> enough about the motivation. Sorry about that.
>
> Every time code is about to access to contents of an optional it needs to
> do one of two things:
>
>    - Have some prior knowledge so it "knows" that the optional has valid
>    contents, or
>    - Verify, locally, that the optional has a valid value prior to access.
>
> If the code does not do one of those two things then that's a setup for a
> crash in the future.
>
> The safest choice of those two, particularly in the face of an evolving
> code base, is to do the local check.
>
> Putting all of the *Keys in ValidatorKeys into a single optional means
> that with one check...
>
> if (validatorKeys_.keys)
>
> I can be assured that it's safe to access any and all of the *Keys in
> ValidatorKeys::Keys.
>
> If the *Keys are stored in three separate optionals, then I have to do
> separate checks any time I access any of the keys.
>
> So, by putting all three *Keys in a single optional it becomes both
>
>    1. Easy and
>    2. Compute efficient
>
> to see whether accessing *any* of the *Keys is safe.
>
> Just in case I'm still not clear, I've made a commit where the optional
> is checked *every time* before the *Keys in ValidatorKeys are accessed.
> That's the top-most commit here:
> https://github.com/scottschurr/rippled/commits/chenna-ppp
> <https://urldefense.com/v3/__https://github.com/scottschurr/rippled/commits/chenna-ppp__;!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcOokG_iuA$>
>
> You're welcome to cherry-pick that commit if you want. But don't blindly
> drop it in. Look at the places the checks are added. I might have messed
> something up.
>
> Does this all make sense? From a larger perspective we're pushing first
> for safety and correctness. Once we have that, then we can make things
> efficient.
> ------------------------------
>
> In src/ripple/app/misc/impl/ValidatorKeys.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1283673587__;Iw!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcPUhHeNFg$>
> :
>
> > @@ -56,9 +56,7 @@ ValidatorKeys::ValidatorKeys(Config const& config, beast::Journal j)
>              }
>              else
>              {
> -                secretKey = token->validationSecret;
> -                publicKey = pk;
> -                masterPublicKey = m->masterKey;
> +                keys.emplace(token->validationSecret, pk, m->masterKey);
>
> Nice use of emplace()!
> ------------------------------
>
> In src/ripple/app/misc/impl/ValidatorKeys.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1283681301__;Iw!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcO8h0vLAw$>
> :
>
> > @@ -83,10 +81,11 @@ ValidatorKeys::ValidatorKeys(Config const& config, beast::Journal j)
>          }
>          else
>          {
> -            secretKey = generateSecretKey(KeyType::secp256k1, *seed);
> -            publicKey = derivePublicKey(KeyType::secp256k1, *secretKey);
> -            masterPublicKey = *publicKey;
> -            nodeID = calcNodeID(*publicKey);
> +            SecretKey sk = generateSecretKey(KeyType::secp256k1, *seed);
> +            PublicKey pk = derivePublicKey(KeyType::secp256k1, sk);
> +            keys.emplace(sk,
>
> Nice use of emplace()!
> ------------------------------
>
> In src/ripple/app/misc/ValidatorKeys.h
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1283684623__;Iw!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcMjB4KzLQ$>
> :
>
> > +            SecretKey const& secret_,
> +            PublicKey const& public_,
> +            PublicKey const& masterPublic_
>
> Please have the order of your constructor arguments match the order of the
> public members in Keys. Putting them in a different order only promotes
> confusion.
> ------------------------------
>
> In src/ripple/app/misc/impl/ValidatorKeys.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1283688029__;Iw!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcNdLicbeg$>
> :
>
> > @@ -83,10 +81,11 @@ ValidatorKeys::ValidatorKeys(Config const& config, beast::Journal j)
>          }
>          else
>          {
> -            secretKey = generateSecretKey(KeyType::secp256k1, *seed);
> -            publicKey = derivePublicKey(KeyType::secp256k1, *secretKey);
> -            masterPublicKey = *publicKey;
> -            nodeID = calcNodeID(*publicKey);
> +            SecretKey sk = generateSecretKey(KeyType::secp256k1, *seed);
> +            PublicKey pk = derivePublicKey(KeyType::secp256k1, sk);
> +            keys.emplace(sk,
> +                         derivePublicKey(KeyType::secp256k1, sk), pk);
>
> Notice that you are calling derivePublicKey() with the same arguments
> twice. Since the arguments are the same the two calls return identical
> results. I think, compute time wise, you're better off if you make the
> emplace() call like this:
>
> keys.emplace(sk, pk, pk);
>
> ------------------------------
>
> In src/ripple/app/consensus/RCLConsensus.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1283705757__;Iw!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcN4EZtMIw$>
> :
>
> > @@ -105,13 +105,13 @@ RCLConsensus::Adaptor::Adaptor(
>          JLOG(j_.info()) << "Validator identity: "
>                          << toBase58(
>                                 TokenType::NodePublic,
> -                               *validatorKeys_.masterPublicKey);
> +                               *validatorKeys_.getMasterPublicKey());
>
> The optional returning getter functions really miss the point of why the
> three keys were gathered together into a single optional. If I'm not
> confident that an optional I'm about to dereference is unseated I need to
> check that optional -- does it have a value or not?
>
> This function is not making that check. But it *could* make that check by
> replacing the code on line 101 with...
>
>     if (validatorKeys_.keys)
>
> If you make that single change they you can be confident that every time
> you dereference any of the keys in validatorKeys_.keys there is a valid
> key to access.
>
> With the added getter function I'm back to not knowing whether the
> returned value is useable. I should check the return value every time.
>
> So I suggest that you remove the getter functions for two reasons:
>
>    - They hide the fact that a single check will let me know that all (or
>    none) of the keys are valid.
>    - Notice that ValidatorKeys has a bunch of public data members.
>    Embrace that. Go ahead and access the public data members as data members.
>
> Does that make sense? It's possible I missed something. I'm open to being
> wrong.
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*pullrequestreview-1561794763__;Iw!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcNoocg2rQ$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AFB4TNM3DILL4JHNDLOY7I3XTQIW3ANCNFSM6AAAAAAZ43VWPQ__;!!PZTMFYE!5M7yWugj9lZGIdANUq_dIy6g1tt-tb9qQGmaXwWXl0guoI4Y-N89ukBKjLpSwGW4UkqJTTR1a7_CbTGAOcOvNq7B8w$>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

ckeshava commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -98,20 +98,20 @@ RCLConsensus::Adaptor::Adaptor(
             std::to_string(valCookie_) + ")";
 
-    if (validatorKeys_.nodeID != beast::zero)
+    if (validatorKeys_.nodeID != beast::zero && validatorKeys_.keys)
```
@scottschurr  Hello Scott, thank you for working on the commit yourself, that was very sweet of you.

I noticed that you didn't include the first condition in your work, I missed that :) thanks

ckeshava replied:
I'll update it along with the forthcoming PR comments, noting it here ^

scottschurr commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -799,10 +808,17 @@ RCLConsensus::Adaptor::validate(
 
+    if(!validatorKeys_.keys)
+    {
+        JLOG(j_.fatal()) << "RCLConsensus::Adaptor::validate: ValidatorKeys "
```
This should become the same logging level as is chosen on line 216.

scottschurr commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -210,12 +210,21 @@ RCLConsensus::Adaptor::propose(RCLCxPeerPos::Proposal const& proposal)
+
+    if(!validatorKeys_.keys)
+    {
+        JLOG(j_.fatal()) << "RCLConsensus::Adaptor::propose: ValidatorKeys "
```
I think that logging this as `fatal()` is overkill.  I suggest `error()` or even `warn()`.  If a bug gets in that triggers this message then the logs will be swamped with this log message.  At `fatal()` level it is hard to remove the message.  And the recovery, simply returning, is benign.

ckeshava replied:
okay

scottschurr reviewed (commented):
Your most recent changes regarding the `optional` in `ValidatorKeys` look good to me.  Nice work.

This is another partial review.  I'm not done yet, but I wanted to get these suggested changes to you before I leave for PTO.

ckeshava commented:
Thank you very much, I'll work on this tomorrow
Regards,
Keshava


On Wed, Aug 9, 2023 at 4:28 PM Scott Schurr ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Your most recent changes regarding the optional in ValidatorKeys look
> good to me. Nice work.
>
> This is another partial review. I'm not done yet, but I wanted to get
> these suggested changes to you before I leave for PTO.
> ------------------------------
>
> In src/ripple/app/main/Application.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1286492055__;Iw!!PZTMFYE!5MegXq10PCh7U2wukkLXLgmRhYWWmYwwcNxX-32hxmC_0V5exwFibIAhFmMxSHsLohRbRzZFy_wqdGcDvvOpcSNKcw$>
> :
>
> > +        // do not use the getter function from ValidatorKeys class because
> +        // const& is returned from this function
>
> I think the comment here no longer applies. The comment should probably be
> removed.
> ------------------------------
>
> In src/ripple/app/main/Application.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1289329172__;Iw!!PZTMFYE!5MegXq10PCh7U2wukkLXLgmRhYWWmYwwcNxX-32hxmC_0V5exwFibIAhFmMxSHsLohRbRzZFy_wqdGcDvvPVz-gwYQ$>
> :
>
> > @@ -584,7 +584,7 @@ class ApplicationImp : public Application, public BasicApp
>          return *m_jobQueue;
>      }
>
> -    std::pair<PublicKey, SecretKey> const&
> +    std::optional<std::pair<PublicKey, SecretKey>> const&
>      nodeIdentity() override
>
> Changing Application::nodeIdentity() to return an optional, is a problem
> for all callers of nodeIdentity(). To safely use the new interface every
> caller needs to check whether the optional is unseated before using it.
> And, by the way, most of the places that were "fixed" to use the new
> interface don't verify that the optional has a valid value. In my opinion
> that's bad programming practice.
>
> So I dug into the question. It turns out that the nodeIdentity is not
> actually optional. It's always available shortly after the Application is
> constructed. So the window where the nodeIdentity is unknown is very
> short. And currently there are no callers of nodeIdentity() before the
> value is known.
>
> This leaves us in a place where we can say that calling nodeIdentity too
> early in the boot process is a programming error. In fact, I think it's a
> good idea to treat it like that.
>
> With that in mind, I suggest that the nodeIdentity implementation be
> changed to look like this:
>
>     std::pair<PublicKey, SecretKey> const&
>     nodeIdentity() override
>     {
>         if (nodeIdentity_)
>             return *nodeIdentity_;
>
>         LogicError(
>             "Accessing Application::nodeIdentity() before it is initialized.");
>     }
>
> A LogicError has the effect of crashing the server with a good error
> message. If the server is early in the boot process then crashing is often
> a better choice that throwing an exception or just writing to the log.
>
> Thoughts?
> ------------------------------
>
> In src/ripple/app/consensus/RCLConsensus.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1289342420__;Iw!!PZTMFYE!5MegXq10PCh7U2wukkLXLgmRhYWWmYwwcNxX-32hxmC_0V5exwFibIAhFmMxSHsLohRbRzZFy_wqdGcDvvPv3HgG0Q$>
> :
>
> > @@ -799,10 +808,17 @@ RCLConsensus::Adaptor::validate(
>          validationTime = lastValidationTime_ + 1s;
>      lastValidationTime_ = validationTime;
>
> +    if(!validatorKeys_.keys)
> +    {
> +        JLOG(j_.fatal()) << "RCLConsensus::Adaptor::validate: ValidatorKeys "
>
> This should become the same logging level as is chosen on line 216.
> ------------------------------
>
> In src/ripple/app/consensus/RCLConsensus.cpp
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*discussion_r1289342450__;Iw!!PZTMFYE!5MegXq10PCh7U2wukkLXLgmRhYWWmYwwcNxX-32hxmC_0V5exwFibIAhFmMxSHsLohRbRzZFy_wqdGcDvvNKI_Wjgg$>
> :
>
> > @@ -210,12 +210,21 @@ RCLConsensus::Adaptor::propose(RCLCxPeerPos::Proposal const& proposal)
>          proposal.prevLedger().begin(), proposal.prevLedger().size());
>      prop.set_proposeseq(proposal.proposeSeq());
>      prop.set_closetime(proposal.closeTime().time_since_epoch().count());
> +
> +    if(!validatorKeys_.keys)
> +    {
> +        JLOG(j_.fatal()) << "RCLConsensus::Adaptor::propose: ValidatorKeys "
>
> I think that logging this as fatal() is overkill. I suggest error() or
> even warn(). If a bug gets in that triggers this message then the logs
> will be swamped with this log message. At fatal() level it is hard to
> remove the message. And the recovery, simply returning, is benign.
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/XRPLF/rippled/pull/4607*pullrequestreview-1566313966__;Iw!!PZTMFYE!5MegXq10PCh7U2wukkLXLgmRhYWWmYwwcNxX-32hxmC_0V5exwFibIAhFmMxSHsLohRbRzZFy_wqdGcDvvPmGwCm4g$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AFB4TNODJSMPQ456V5VF7X3XUQMILANCNFSM6AAAAAAZ43VWPQ__;!!PZTMFYE!5MegXq10PCh7U2wukkLXLgmRhYWWmYwwcNxX-32hxmC_0V5exwFibIAhFmMxSHsLohRbRzZFy_wqdGcDvvNt90aLyg$>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

ckeshava commented on src/ripple/app/main/Application.cpp:
```diff
@@ -593,7 +593,12 @@ class ApplicationImp : public Application, public BasicApp
+        if (!validatorKeys_.keys)
+            return PublicKey::emptyPubKey;
+
+        // do not use the getter function from ValidatorKeys class because
+        // const& is returned from this function
```
ok

ckeshava commented on src/ripple/app/main/Application.cpp:
```diff
@@ -584,7 +584,7 @@ class ApplicationImp : public Application, public BasicApp
 
-    std::pair<PublicKey, SecretKey> const&
+    std::optional<std::pair<PublicKey, SecretKey>> const&
     nodeIdentity() override
```
thanks for catching this!

ckeshava commented:
thanks @scottschurr  I have address your recent comments

ximinez reviewed (changes requested):
Partial review...

ckeshava commented on src/ripple/protocol/PublicKey.h:
```diff
@@ -59,13 +59,38 @@ namespace ripple {
+    // publisher's public key and hence needs to be mapped with the emptyPubKey
+    static const PublicKey emptyPubKey;
+
+    static PublicKey getEmptyPubKey()
```
Okay, but I need `emptyPubKey` to be `public` for its utility across the codebase. I don't know if I can make that variable `private` :/

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -162,6 +163,8 @@ ValidatorList::load(
 
         auto const ret = strUnHex(key);
 
+        // The check for publicKeyType(...) ensures that default-constructed
+        // public keys are not passed here.
```
You are correct, I'd still retain the check on the publicKeyType for additional security

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -231,13 +234,10 @@ ValidatorList::load(
         // Config listed keys never expire
-        auto& current = it->second.current;
-        if (inserted)
-            current.validUntil = TimeKeeper::time_point::max();
+        auto& current = localPublisherList;
+        current.validUntil = TimeKeeper::time_point::max();
         current.list.emplace_back(*id);
-        it->second.status = PublisherStatus::available;
         ++count;
```
Ok thanks. As far as setting `validUntil` only once, I don't think the above code helps. The if-condition is triggered for every positive value of `count`, but I got your intent behind the suggestion 👍

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -195,8 +198,8 @@ ValidatorList::load(
     // Treat local validator key as though it was listed in the config
-    if (localPubKey_.size())
-        keyListings_.insert({localPubKey_, 1});
+    if (localPubKey_ && *localPubKey_ != PublicKey::emptyPubKey)
```
Several of the unit tests do not specify `ValidatorKeys` during the setup of the Application object. In the existing codebase, such keys are being default-constructed. 

In the spirit of retaining old behavior, I have written this failsafe: https://github.com/XRPLF/rippled/blob/2af13a7ad5b8e10e6fd230aebfba4819fd3a453b/src/ripple/app/main/Application.cpp#L1333

How can I make this better? Do I need to update all the unit tests? 😓

ckeshava commented on src/ripple/app/misc/ValidatorKeys.h:
```diff
@@ -36,9 +36,30 @@ class Config;
+
+        Keys() = delete;
+        Keys(
+            SecretKey const& secret_,
+            PublicKey const& public_,
+            PublicKey const& masterPublic_
```
thanks for catching it, I missed it!

ckeshava commented:
@ximinez  thanks for your comments, I have addressed them in 3bd70664ab94236c015f2876cd3c4dbc362e45e5
(except for the one about the assert statement)

ckeshava commented on src/ripple/app/misc/impl/ValidatorKeys.cpp:
```diff
@@ -56,9 +56,7 @@ ValidatorKeys::ValidatorKeys(Config const& config, beast::Journal j)
-                secretKey = token->validationSecret;
-                publicKey = pk;
-                masterPublicKey = m->masterKey;
+                keys.emplace(token->validationSecret, pk, m->masterKey);
```
thanks!!

ximinez commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -213,13 +213,22 @@ RCLConsensus::Adaptor::propose(RCLCxPeerPos::Proposal const& proposal)
     prop.set_proposeseq(proposal.proposeSeq());
     prop.set_closetime(proposal.closeTime().time_since_epoch().count());
+
+    if(!validatorKeys_.keys)
+    {
+        JLOG(j_.warn()) << "RCLConsensus::Adaptor::propose: ValidatorKeys "
+                            "not set: \n";
+        return;
+    }
```
Since you know that `keys` is set at this point, it might make sense, improve readability a little, and maybe speed things up by an amount too small to measure to set a reference and use it for the rest of the function:
```
auto const& keys = *validatorKeys_.keys;
```

ximinez commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -812,10 +821,17 @@ RCLConsensus::Adaptor::validate(
         validationTime = lastValidationTime_ + 1s;
     lastValidationTime_ = validationTime;
 
+    if(!validatorKeys_.keys)
+    {
+        JLOG(j_.warn()) << "RCLConsensus::Adaptor::validate: ValidatorKeys "
+                            "not set\n";
+        return;
+    }
```
As above, since you know that `keys` is set at this point, it might make sense, improve readability a little, and maybe speed things up by an amount too small to measure to set a reference and use it for the rest of the function:
```
auto const& keys = *validatorKeys_.keys;
```

ximinez commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -991,7 +1007,8 @@ RCLConsensus::Adaptor::preStartRound(
 {
     // We have a key, we do not want out of sync validations after a restart
     // and are not amendment blocked.
-    validating_ = validatorKeys_.publicKey.size() != 0 &&
+    validating_ = validatorKeys_.keys.has_value() &&
+        validatorKeys_.keys->publicKey != PublicKey::emptyPubKey &&
```
It should be impossible for `keys` to be initialized with `PublicKey::emptyKey`. The check is still good, but I suggest adding an assert before this line:
```
assert(!validatorKeys_.keys || validatorKeys_.keys->publicKey != PublicKey::emptyPubKey);
```
Also there's not much need for `has_value()` unless you're dealing with ambiguity.

ximinez reviewed (changes requested):
Partial review

ckeshava commented on src/ripple/app/misc/impl/ValidatorKeys.cpp:
```diff
@@ -56,9 +56,7 @@ ValidatorKeys::ValidatorKeys(Config const& config, beast::Journal j)
-                secretKey = token->validationSecret;
-                publicKey = pk;
-                masterPublicKey = m->masterKey;
+                keys.emplace(token->validationSecret, pk, m->masterKey);
```
hmm, I have fixed this and the below comment in my local machine. The build is completing correctly on my local machine too

ckeshava commented on src/ripple/app/misc/impl/ValidatorKeys.cpp:
```diff
@@ -83,10 +81,11 @@ ValidatorKeys::ValidatorKeys(Config const& config, beast::Journal j)
+            SecretKey sk = generateSecretKey(KeyType::secp256k1, *seed);
+            PublicKey pk = derivePublicKey(KeyType::secp256k1, sk);
+            keys.emplace(sk,
+                         derivePublicKey(KeyType::secp256k1, sk), pk);
```
thanks! 👍

ckeshava commented:
thanks @ximinez  I have addressed the recent batch of comments 👍

ximinez reviewed (changes requested):
This does not include your latest commit

ckeshava commented on src/ripple/app/main/Application.cpp:
```diff
@@ -1311,9 +1318,25 @@ ApplicationImp::setup(boost::program_options::variables_map const& cmdline)
+
+            PublicKey localSigningKey = PublicKey::emptyPubKey;
+            if (validatorKeys_.keys)
+                localSigningKey = validatorKeys_.keys->publicKey;
```
hmm, I have fixed this in my most recent commit... I wonder why it's not showing up :/

ximinez replied:
I probably wrote this comment before loading that last commit. It's there now.

ximinez commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -1317,14 +1318,16 @@ ValidatorList::verify(
     if (!sig ||
         !ripple::verify(
-            publisherManifests_.getSigningKey(masterPubKey),
+            *publisherManifests_.getSigningKey(masterPubKey),
```
Addendum: assertion or not, you need to explicitly check that `signingKey` is set before dereferencing it.

```
    auto const signingKey = publisherManifests_.getSigningKey(masterPubKey);
    assert(revoked || signingKey);

    if (revoked || !signingKey || result == ManifestDisposition::invalid)
        return {ListDisposition::untrusted, masterPubKey};

    auto const sig = strUnHex(signature);
    auto const data = base64_decode(blob);
    if (!sig ||
        !ripple::verify(
            *signingKey,
[...]
```

ckeshava replied:
okay, I'll make the change about using a local variable. But the `assert(revoked || signingKey)` implies that a non-revoked manifest must contain a signingKey. Hence I felt it was not necessary to check the signingKey again, but I'll make the change 👍

ximinez commented on src/ripple/app/misc/ValidatorList.h:
```diff
@@ -343,7 +343,7 @@ class ValidatorList
     bool
     load(
-        PublicKey const& localSigningKey,
+        std::optional<PublicKey const> localSigningKey,
```
Should be ` std::optional<PublicKey> const&`

ckeshava replied:
ok. I get that the `const&` helps avoid the extra copy. Shouldn't it be `std::optional<PublicKey const> const&` to ensure that the contents of the optional is not modified?

I don't know how the optional works, but if it stores a pointer to the PublicKey object, then your suggestion could still violate the `const-` ness. On the other hand, if an entire copy of the `PublicKey` object is stored within the optional, then your version works just fine.

ximinez replied:
According to https://en.cppreference.com/w/cpp/utility/optional/operator*, the `const` overloads of the `->` and `*` operators return a const ref. The same is true for https://en.cppreference.com/w/cpp/utility/optional/value. So it would not be easy to modify the value. (I almost never say "impossible", because you can always do `const_cast` craziness and such.)

ckeshava replied:
okay, I'll go through these references, thanks

ckeshava commented on src/ripple/app/misc/impl/ValidatorKeys.cpp:
```diff
@@ -83,10 +81,11 @@ ValidatorKeys::ValidatorKeys(Config const& config, beast::Journal j)
+            PublicKey pk = derivePublicKey(KeyType::secp256k1, sk);
+            keys.emplace(pk,
+                         derivePublicKey(KeyType::secp256k1, sk), sk);
+            nodeID = calcNodeID(pk);
```
ok, I will add the const-qualifier in the next commit. I have removed the second function call to `derivePublicKey`. I don't think you are looking at the latest version of my updates.

ckeshava commented on src/test/rpc/KeyGeneration_test.cpp:
```diff
@@ -388,8 +386,8 @@ class WalletPropose_test : public ripple::TestSuite
             BEAST_EXPECT(!contains_error(error));
-            BEAST_EXPECT(ret.first.size() != 0);
-            BEAST_EXPECT(ret.first == publicKey);
+            BEAST_EXPECT(ret->first.size() != 0);
+            BEAST_EXPECT(ret->first == publicKey);
```
I think `BEAST_EXPECT(ret);` is a better way to communicate that `ret` must not be unseated. What do you think?

ximinez replied:
Better than what? My suggestion includes `BEAST_EXPECT(ret)`, and does communicate that it must be seated. If you're saying it would be better as a standalone function call, then I disagree. If `ret` is somehow unseated, then having the `if` statement will still cause the test to fail, but the suite will continue. If it's standalone, and `ret` is somehow unseated, then the next test, which dereferences `ret`, will `assert`, crash, or have UB.

ckeshava replied:
But all the `ret`'s are in separate scopes. Even if one of the `ret` is unseated, I don't believe it crashes the subsequent tests.

Depending on the nature of the test case, only `if` or the `else` portion of the code is relevant. For example, this [test](https://github.com/ckeshava/rippled/blob/7df60ffd858d0f0f65aca8d5f5ff972887bf4f2c/src/test/rpc/KeyGeneration_test.cpp#L708) must necessarily have an unseated `ret`.

This [test](https://github.com/ckeshava/rippled/blob/7df60ffd858d0f0f65aca8d5f5ff972887bf4f2c/src/test/rpc/KeyGeneration_test.cpp#L389) must have a valid value inside `ret` for correctness. What should I write in the else portion of these tests?

ximinez replied:
> But all the `ret`'s are in separate scopes. Even if one of the `ret` is unseated, I don't believe it crashes the subsequent tests.

It asserts in Windows builds. That counts as a crash.

https://en.cppreference.com/w/cpp/utility/optional/operator*
"The behavior is undefined if `*this` does not contain a value. "

Potential UB is unacceptable except in the rarest of circumstances, and this isn't one of them.

> This [test](https://github.com/ckeshava/rippled/blob/7df60ffd858d0f0f65aca8d5f5ff972887bf4f2c/src/test/rpc/KeyGeneration_test.cpp#L389) must have a valid value inside `ret` for correctness. What should I write in the else portion of these tests?

Nothing. If the `ret` is unseated, there's nothing else to check. If you use the `if(BEAST_EXPECT(ret))` construct, the suite will be marked as failed, and you'll get that result at the end of the run.

ckeshava replied:
ok, this makes it clear, thanks. I'll update the code

ckeshava replied:
hello Ed, I have addressed these comments in f69b0d0bdf62d6aa7cfb182decbc3ded350a3422

ckeshava replied:
I believe the negative tests do not need any other modification (`BEAST_EXPECT(!ret);` cases) because there is no opportunity for undefined behavior.

ckeshava commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -214,22 +216,21 @@ RCLConsensus::Adaptor::propose(RCLCxPeerPos::Proposal const& proposal)
     prop.set_nodepubkey(
-        validatorKeys_.keys->publicKey.data(), validatorKeys_
-                                                   .keys->publicKey.size());
+        keys.publicKey.data(), validatorKeys_.keys->publicKey.size());
```
thanks! i missed it

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -1669,8 +1672,9 @@ ValidatorList::getJson() const
         if (it != keyListings_.end())
         {
+            assert(manifest.revoked() || manifest.signingKey);
             jSigningKeys[toBase58(TokenType::NodePublic, manifest.masterKey)] =
-                toBase58(TokenType::NodePublic, manifest.signingKey);
+                toBase58(TokenType::NodePublic, *manifest.signingKey);
```
oh good catch! I have placed a check to ensure that the signingKey is seated before it's use. Let me know if this is ok.

ckeshava commented:
@ximinez  thanks a lot for your detailed comments. I have addressed them in afa642506a5abc8896aff8f85ea260fd0971f96c

ximinez reviewed (changes requested):
1. This branch doesn't build in the current state.
2. Undo the changes you made under `secp256k1`. That's external code, and shouldn't generally be touched unless updating versions.
3. The branch is pretty badly out of date. Please merge in the latest `develop`.

ximinez commented:
I notice the debug builds are still failing.

ckeshava commented:
@ximinez  I've updated the unit tests to fix the failing builds. I'm surprised that the other platforms did not catch these failures!

ximinez commented on src/ripple/protocol/PublicKey.h:
```diff
@@ -61,13 +61,41 @@ namespace ripple {
-    std::uint8_t buf_[33];  // should be large enough
+    // All the constructed public keys are valid, non-empty and contain 33
+    // bytes of data.
+    static constexpr std::size_t size_ = 33;
```
Now that `size_` is a `constexpr`, any checks of it don't make sense. I've already pointed out issues in some of the ctors, but I noticed some more.
* It's impossible for `empty()` to return `true`. It should be removed entirely.
* Any remaining uses of `size_` should be checked to ensure they make sense. I don't think there are many.

ximinez commented on src/ripple/protocol/PublicKey.h:
```diff
@@ -61,13 +61,41 @@ namespace ripple {
     using const_iterator = std::uint8_t const*;
 
-    PublicKey() = default;
+    // This default constructed public key is used only in ValidatorList class.
+    // Publishers' master PublicKeys are used to map into their respective UNLs
+    // emptyPubKey is used to represent the keys specified in the local config
+    // file. Unlike the other published UNL, this list is not signed with any
+    // publisher's public key and hence needs to be mapped with the emptyPubKey
+    static const PublicKey emptyPubKey;
```
This comment is not true any more. The local config is now represented by its own variable (`localPublisherList`), and isn't stored in `publisherLists_`. I don't think `emptyPubKey` is used anywhere "useful".

I say get rid of it entirely. Any place returning it as a placeholder or fallback is _wrong_, and needs to use `std::optional` instead. All the checks become unnecessary, unless they're checking for a seated `optional`.

ckeshava commented on src/test/protocol/PublicKey_test.cpp:
```diff
@@ -363,7 +363,8 @@ class PublicKey_test : public beast::unit_test::suite
 
         // Try some random secret keys
-        std::array<PublicKey, 32> keys;
+        std::vector<PublicKey> keys;
+        keys.reserve(32);
 
         for (std::size_t i = 0; i != keys.size(); ++i)
             keys[i] = derivePublicKey(keyType, randomSecretKey());
```
thanks for letting me know! I assumed that `reserve` would ensure that `size()` returns the capacity of the vector, my bad :/

ckeshava commented on src/test/app/LedgerLoad_test.cpp:
```diff
@@ -63,7 +63,7 @@ class LedgerLoad_test : public beast::unit_test::suite
 
         Env env{*this};
-        Account prev;
+        Account prev("previous_account");
```
About 10 lines below, `prev` is used in `env.trust()` and `env(pay(...))` commands. I don't think `prev` can be used as a recipient account if its an unseated optional

ximinez replied:
Yes, but those transactions are skipped on the first run through the loop. `prev` is assigned at the end of the loop. So it's guaranteed that `prev` will have a value when it is dereferenced.

```
        std::optional<Account> prev;

        for (auto i = 0; i < 20; ++i)
        {
            Account acct{"A" + std::to_string(i)};
            env.fund(XRP(10000), acct);
            env.close();
            if (i > 0 && BEAST_EXPECT(prev))
            {
                env.trust(acct["USD"](1000), *prev);
                env(pay(acct, *prev, acct["USD"](5)));
            }
            env(offer(acct, XRP(100), acct["USD"](1)));
            env.close();
            prev.emplace(std::move(acct));
        }
```

Another, simpler option is instead of initializing it with a dummy value that's going to get thrown away, initialize it with a value you know it will have. At least then you know it will be memoized.
```
Account prev("A19");
```

ckeshava replied:
thanks Ed, this is very thoughtful!

ckeshava replied:
Hmm, that's an interesting thought. Surprisingly, the `operator=` code generates 10-20 lines more assembly than the `emplace` code. I'm guessing there are other aspects of performance to consider. https://godbolt.org/z/hW1a4GroP

ckeshava commented on src/ripple/protocol/PublicKey.h:
```diff
@@ -59,13 +59,38 @@ namespace ripple {
+        std::array<unsigned char, 33> buf;
+
+        // set the values to zero to simulate a default-constructed Public Key
+        for(unsigned int i = 0; i < 33; i++)
```
I've removed this function

ckeshava commented on src/ripple/protocol/impl/PublicKey.cpp:
```diff
@@ -178,11 +179,11 @@ PublicKey::PublicKey(Slice const& slice)
         LogicError("PublicKey::PublicKey invalid type");
-    size_ = slice.size();
+//    size_ = slice.size();
     std::memcpy(buf_, slice.data(), size_);
```
thanks! great point

ckeshava commented on src/ripple/rpc/impl/TransactionSign.cpp:
```diff
@@ -369,10 +369,14 @@ transactionPreProcessImpl(
     Json::Value jvResult;
-    auto const [pk, sk] = keypairForSignature(params, jvResult);
-    if (contains_error(jvResult))
+    std::optional<std::pair<PublicKey, SecretKey>> keyPair =
+        keypairForSignature(params, jvResult);
+    if (!keyPair || contains_error(jvResult))
         return jvResult;
 
+    PublicKey pk = keyPair->first;
+    SecretKey sk = keyPair->second;
```
You're right. I think the existing check is insufficient. If the `keyPair` is unseated and the `jss::error` field is unseated, then the `if` condition is not triggered and we're in a pickle

ckeshava commented on src/ripple/rpc/impl/TransactionSign.cpp:
```diff
@@ -977,7 +981,9 @@ transactionSignFor(
 
     // Add and amend fields based on the transaction type.
     Buffer multiSignature;
-    PublicKey multiSignPubKey;
+    PublicKey multiSignPubKey = derivePublicKey(
+        KeyType::secp256k1,
+        generateSecretKey(KeyType::secp256k1, generateSeed("dummyPassPhrase")));
     SigningForParams signForParams(
         *signerAccountID, multiSignPubKey, multiSignature);
```
Thanks for your work Ed, that helps 👍 
I made a slight modification to your commit [here](https://github.com/ckeshava/rippled/tree/refactorSignParams), what do you think of this?

I believe `transactionPreProcessImpl` should fill and return the `SiningForParams` object. This localizes the writes to this object and improves readability (but the `JSONRPC` unit tests are failing, I'm working on that).

ckeshava commented:
@ximinez I have addressed your comments, thanks for your suggestions.
Apologies for missing the earlier review, it slipped my mind.

ckeshava commented:
@ximinez @scottschurr Hello, thank you for looking into this PR. I believe I have addressed all the comments, please review it at your leisure. thanks!

ximinez commented on src/ripple/app/misc/impl/Manifest.cpp:
```diff
@@ -139,14 +141,17 @@ deserializeManifest(Slice s, beast::Journal journal)
         }
 
-        return m;
+        // If the signingKey has not been specified, then the constructor
+        // of the Manifest class uses the masterKey as a fail-safe
+        // signingKey.
+        return Manifest(serialized, masterKey, signingKey, seq, domain);
```
This comment is no longer accurate.

ximinez commented on src/ripple/app/misc/impl/Manifest.cpp:
```diff
@@ -496,10 +512,11 @@ ManifestCache::applyManifest(Manifest m)
             iter->second.sequence);
 
-    signingToMasterKeys_.erase(iter->second.signingKey);
+    assert(iter->second.signingKey || iter->second.revoked());
+    signingToMasterKeys_.erase(*iter->second.signingKey);
```
It should be impossible for a revoked key to get to this point, so the assert should be `assert(iter->second.signingKey && !iter->second.revoked())`.

ckeshava replied:
I'm trying to better understand this piece of code. I found that `assert(m == iter->second)` fails. Specifically, `m` has a higher sequence number than the manifest in `iter->second`. Isn't that wrong? I was of the understanding that `map_` (parent data structure of `iter`) needs to be identical to the input of the function (`m`)

ximinez replied:
`iter->second` is holding the old / existing manifest. `m` is the new manifest. They're for the same `masterKey`, but they're expected to be different. In fact, `prewriteCheck` checks that.

You could assert that 
```
assert(
        iter->second.masterKey == m.masterKey &&
        iter->second.sequence < m.sequence);
```
but that's redundant.

ximinez commented on src/ripple/overlay/impl/PeerImp.h:
```diff
@@ -339,7 +339,7 @@ class PeerImp : public Peer,
     checkTracking(std::uint32_t seq1, std::uint32_t seq2);
 
-    PublicKey const&
+    PublicKey
     getNodePublic() const override
     {
         return publicKey_;
```
I really don't like this. It looks like you got rid of the `const&` because you need to generate a `PublicKey` on the fly in the `TestPeer` test class. Instead of forcing an unnecessary copy to accommodate a test, change the test class to store that generated key in a class member variable, then return the variable.

ximinez commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -30,24 +30,35 @@ Json::Value
 
-    // pk is maskter key, eg no ephemeral key, eg no manifest, just return
-    if (mk == pk)
+    // validationPK is maskter key, eg no ephemeral key, eg no manifest, just
+    // return
+    if (mk == validationPK)
         return ret;
```
Typo: `maskter` -> `master`

ckeshava commented on src/test/app/LedgerLoad_test.cpp:
```diff
@@ -63,21 +63,21 @@ class LedgerLoad_test : public beast::unit_test::suite
             env.fund(XRP(10000), acct);
             env.close();
-            if (i > 0)
+            if (i > 0 && BEAST_EXPECT(prev))
```
this piece of code was one of your suggestions from a while ago :)

ximinez reviewed (approved):
Looks good!

ckeshava commented:
thanks for your feedback @ximinez !

ckeshava commented:
hello @scottschurr , would you be able to review this PR? thanks!

intelliot commented:
note: this is on Scott S's to-review list

scottschurr commented on src/ripple/app/misc/impl/Manifest.cpp:
```diff
@@ -421,11 +433,12 @@ ManifestCache::applyManifest(Manifest m)
             return ManifestDisposition::badMasterKey;
         }
 
+        assert(m.signingKey || m.revoked());
```
We're in a function that returns the disposition of the manifest.  I think that a better solution than a `assert` would be to actually perform the check and return the appropriate response.  Perhaps something like this inside of the immediately following condition...
```
        if (!revoked)
        {
            if (!m.signingKey)
            {
                JLOG(j_.warn()) << to_string(m)
                                << ": is not revoked but has no signing key";
                return ManifestDisposition::invalid;
            }

            // Sanity check: the ephemeral key of this manifest should not be
            ...
```

scottschurr reviewed (commented):
Changes look pretty good to me.  You and @ximinez have put a lot of work into this.

I left a few suggestions for changes.  Consider them and let me know what you think.  Thanks.

ckeshava commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -2456,10 +2456,11 @@ NetworkOPsImp::getServerInfo(bool human, bool admin, bool counters)
-        if (!app_.getValidationPublicKey().empty())
+        if (app_.getValidationPublicKey())
         {
+            assert(app_.validators().localPublicKey());
```
thanks scott, I'll incorporate this suggestion 👍 

I'm confident that a non-empty return value from `getValidationPublicKey()` would imply a non-empty `ValidatorList::localPubKey_`. The `ValidatorList::load()` function is responsible for setting the latter value, and it happens to be only invoked during ApplicationImp's setup: https://github.com/ckeshava/rippled/blob/1a8614895ba7d962b16db990b1028e7146614093/src/ripple/app/main/Application.cpp#L1487

having said that, it took me a considerable amount of time to search through this code, hence your suggestion is helpful.

intelliot commented:
- awaiting update or reply from @ckeshava

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -883,9 +885,11 @@ ValidatorList::applyListsAndBroadcast(
         bool good = true;
-        for (auto const& [pubKey, listCollection] : publisherLists_)
+
+        // localPublisherList is always available. No need to perform the
+        // below check
```
`localPublisherList` contains the information from config file's `SECTION_VALIDATORS`. This is handled in the third (and last portion) of the `ValidatorList::load()` function.

I'm given to understand that the validator information in the local config file does not expire and is always available. Consequently, it is not subject to stringent validity checks.

Note, in the current codebase, such ValidatorLists are mapped to the default constructed Public Key.
I think the comment is correct, perhaps it could be paraphrased to be more informative.

scottschurr replied:
> If the local configuration file does not specify a [validators] section, then localPublisherList.list.empty() == true. This maintains the invariant that all the locally configured PublisherList's are perennially available.
>
> I hope this clears things up

Okay.  Now I think I get what's bothering me about this comment.  Bear with me.  You'll find this code around line 1870 of ValidatorList.cpp:
```
        // Remove if expired
        // ValidatorLists specified in the local config file never expire.
        // Hence, the below steps are not relevant for localPublisherList
        if (collection.status == PublisherStatus::available &&
            collection.current.validUntil <= closeTime)
        {
            removePublisherList(lock, pubKey, PublisherStatus::expired);
            ops.setUNLBlocked();
        }
```
What this code shows to me is that there may be validator lists that expire.  I agree with you that a `localPublisherList` cannot expire.  But other validator lists can.

In light of this, your comment says ...
```
        // localPublisherList is always available. No need to perform the
        // below check
```
To me, at least, the closing sentence seems definitive: `No need to perform the below check`.  I think your intention is that there's no need to perform the below check _on the `localPublisherList`_.

I think I would not have been baffled by the comment if it read, for example, like this:
```
        // localPublisherList never expires, so localPublisherList is excluded
        // from the below check.
```
Am I getting closer?  Or am I still misunderstanding something?

scottschurr replied:
@ckeshava, thoughts on this comment?  I still think I would have understood your intent better with a comment like this:
```
        // localPublisherList never expires, so localPublisherList is excluded
        // from the below check.
```
But that quibble is not enough reason to delay this pull request further.

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -899,12 +903,17 @@ ValidatorList::applyListsAndBroadcast(
     {
-        auto const& pubCollection = publisherLists_[*result.publisherKey];
         assert(
             result.status <= PublisherStatus::expired && result.publisherKey &&
-            pubCollection.maxSequence);
+            publisherLists_[*result.publisherKey].maxSequence);
```
yes, you are right 👍  thanks

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -1452,7 +1476,7 @@ ValidatorList::removePublisherList(
 ValidatorList::count(ValidatorList::shared_lock const&) const
 {
-    return publisherLists_.size();
+    return publisherLists_.size() + (localPublisherList.list.size() > 0);
```
No, this was not influenced by the unit tests.

The locally configured validator lists are mapped to a default-constructed Public Key in the `publisherLists_` map. Since we are removing such default-constructed public keys, I had to place that validator list in a separate variable.

Consequently, the validator list collection has been split into two mutually exclusive parts (the one from the local config file `localPublisherList` and the ones from other peers, which is stored in `publisherLists_` map)

ckeshava commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -1493,6 +1518,20 @@ ValidatorList::expires(ValidatorList::shared_lock const&) const
             res = chainedExpiration;
         }
     }
+
+    if (localPublisherList.list.size() > 0)
+    {
+        PublisherList collection = localPublisherList;
+        // Unfetched
+        auto const& current = collection;
+        auto chainedExpiration = current.validUntil;
+
+        // Earliest
+        if (!res || chainedExpiration < *res)
+        {
+            res = chainedExpiration;
+        }
+    }
```
I have explained the motivation for `localPublisherList ` in a previous comment, let me know if it's clear.

ckeshava commented on src/ripple/protocol/impl/PublicKey.cpp:
```diff
@@ -176,26 +175,27 @@ ed25519Canonical(Slice const& sig)
 
 PublicKey::PublicKey(Slice const& slice)
 {
+    if (slice.size() < size_)
+        LogicError(
+            "PublicKey::PublicKey - Input slice cannot be an undersized "
+            "buffer");
```
thanks to Ed :) @ximinez

ckeshava commented on src/ripple/rpc/handlers/ValidatorInfo.cpp:
```diff
@@ -30,24 +30,28 @@ Json::Value
     auto const ek = context.app.validatorManifests().getSigningKey(mk);
-    ret[jss::ephemeral_key] = toBase58(TokenType::NodePublic, ek);
+
+    assert(ek == validationPK);
```
thanks for this observation.
I was mimicking the structure of the current code base and didn't see the larger picture. I have removed this piece of code.

There is a significant overlap in the information stored between `validatorManifests`, `ValidatorKeys` and `ValidatorList`. It would be very helpful if I could refactor these classes to use a single source of truth, instead of duplicate values smattered around.

but that would be a battle for another PR

ckeshava commented:
@scottschurr I have addressed your comments

scottschurr reviewed (commented):
Looks very good.  I just have a couple of lingering questions.  Thanks for your patience.

ckeshava commented on src/ripple/app/misc/impl/Manifest.cpp:
```diff
@@ -423,9 +442,18 @@ ManifestCache::applyManifest(Manifest m)
+                JLOG(j_.warn()) << to_string(m)
+                                << ": is not revoked and the manifest has no "
+                                   "signing key. Hence, the manifest is "
+                                   "invalid\n";
```
I don't have a strong preference, I can remove it.
I like it better when the log lines are disparate and non-obfuscated.

I don't see any harm in using it.

ckeshava replied:
ok

ckeshava commented:
@scottschurr No, you are correct in saying that all the `PublisherList`'s need to be available for `good == true`, including the `localPublisherList`.

The `localPublisherList` is set at https://github.com/ckeshava/rippled/blob/8aada82e2f7e1c3d37e79b3310eead77fe906cac/src/ripple/app/misc/impl/ValidatorList.cpp#L236
Note, the `validUntil` field is set to a very large value at line 244. (This is conditional on the non-emptiness of the local publisher list.)

If the local configuration file does not specify a `[validators]` section, then `localPublisherList.list.empty() == true`. This maintains the invariant that all the locally configured PublisherList's are perennially available.

I hope this clears things up

ximinez commented on src/ripple/app/misc/impl/ValidatorList.cpp:
```diff
@@ -904,16 +904,12 @@ ValidatorList::applyListsAndBroadcast(
 
-        // The below function call is executed for PublicKeys which are not
-        // specified in the config file. The assert statement above filters
-        // out PublisherLists which are not associated with a master key.
```
I suggest a comment just before or after the `if` that's something like:
```
// broadcastBlobs is only called for PublicKeys which are not specified in the config file.
```

ckeshava replied:
ok

ximinez commented on src/test/overlay/reduce_relay_test.cpp:
```diff
@@ -1058,7 +1058,7 @@ class reduce_relay_test : public beast::unit_test::suite
                 if (event.state_ == State::On)
                 {
-                    assert(event.key_);
+                    BEAST_EXPECT(event.key_);
```
Let's use our old friend `if (BEAST_EXPECT(event.key_))` here, so that it doesn't get dereferenced on the off chance that it's not set.

ckeshava commented on src/ripple/app/misc/NetworkOPs.cpp:
```diff
@@ -2458,9 +2458,11 @@ NetworkOPsImp::getServerInfo(bool human, bool admin, bool counters)
+                info[jss::pubkey_validator] =
+                    toBase58(TokenType::NodePublic, localPubKey.value());
+            }
         }
```
ok.

ckeshava commented:
addressed all of your comments

ximinez reviewed (approved):
I think we're there.

intelliot commented:
@scottschurr and/or @ckeshava : for feedback that has been addressed, feel free to use the `Resolve conversation` button. 

I think `Option`+click (or `Alt`+click on Windows/Linux) on `Show resolved` will expand/collapse all resolved conversations.

scottschurr reviewed (approved):
I have one lingering question about a comment, but that's not sufficient reason to hold up this pull request.

ckeshava commented:
@scottschurr  I have incorporated your suggestions for the comment in `ValidatorList.cpp` file. thanks

scottschurr reviewed (approved):
:+1: Looks great!

ximinez commented:
Suggested commit message:
```
Remove default ctors from SecretKey and PublicKey: (#4607)

* It is now an invariant that all constructed Public Keys are valid, non-empty and contain 33 bytes of data.
* Additionally, the memory footprint of the PublicKey class is reduced. The size_ data member is declared as static.
* Fixes #2942
```
@ckeshava This PR has been through a lot of changes. Feel free to provide additional details that you think might be important.

ckeshava commented:
- Distinguish and identify the `PublisherList` retrieved from the local config file, versus the ones obtained from other validators. 

I'd add this line in the commit message ^^

@ximinez  thanks for help, I couldn't have completed this PR without your contributions
@scottschurr  thanks for your comments scott, you helped point out important mistakes in the PR!
